docs/dev-process/coding-guidelines.md

```diff
@@ -1,92 +1,172 @@
 
-# C# Coding Style
+# C# Coding Guidelines
 
 ## Coding Conventions
 
 As a general rule, our coding convention is to follow the style of the surrounding code.
-Avoid reformatting any code when submitting a PR as it obscures the functional changes of your change.
-We run the [.NET code formatter tool](https://github.com/dotnet/codeformatter) regularly help keep consistent formatting.
+So if a file happens to differ in style from conventions defined here
+(e.g. private members are named `m_member` rather than `_member`),
+the existing style in that file takes precedence.
 
-A basic rule of formatting is to use "Visual Studio defaults".
-Here are some general guidelines
+When making changes, you may find some existing code goes against the conventions defined here.
+In such cases, please avoid reformatting any existing code when submitting a PR as it obscures the functional changes of the PR.
+A separate PR should be submitted for style-only changes.
+We also run the [.NET code formatter tool](https://github.com/dotnet/codeformatter) regularly to keep consistent formatting.
 
-* No tabs, indent 4 spaces.
-* Braces usually go on their own line,
+### Naming Conventions
+
+* Use meaningful, descriptive words for names.
+  For method names, it's encouraged to use `Verb-Object` pair such as **`LoadModule`**.
+
+* Use `_camelCase` to name internal and private fields and use `readonly` where possible.
+  Prefix instance fields with `_`, static fields with `s_` and thread static fields with `t_`.
+  When used on static fields, `readonly` should come after `static` (i.e. `static readonly` not `readonly static`).
+
+* Use `camelCase` to name non-constant local variables.
+
+* Use `PascalCase` to name constant local variables and fields.
+  The only exception is for interop code where the constant should exactly match the name and value of the code you are calling via interop (i.e. `const int ERROR_SUCCESS = 0`).
+
+* Use `PascalCase` to name types and all other type members.
+
+### Layout Conventions
+
+* Use four spaces of indentation (no tabs).
+
+* Avoid more than one blank empty line at any time.
+
+* Avoid unnecessary trailing spaces at the end of a line.
+
+* Braces usually go on their own lines,
   with the exception of single line statements that are properly indented.
-* Use `_camelCase` for instance fields,
-  use `readonly` where possible.
+
+* Namespace imports should be specified at the top of the file,
+  outside of `namespace` declarations and should be sorted alphabetically.
+
+* Fields should be specified at the top within type declarations.
+
+* File encoding should be `ASCII` (preferred) or `UTF8` (with `BOM`) if absolutely necessary.
```

Comment:
Maybe a note to explicitly avoid UTF-16, and a warning that some Windows editors may default to this.

Reply:
I'd like to disallow all `BOM` encodings - tests that need a `BOM` are better off generating the file anyway.

Reply:
Fixed -- Mention that all `bom` encoding should be avoided.